humblegen/tests/rust/service-authorization-using-interceptor/main.rs

```diff
@@ -0,0 +1,67 @@
+mod protocol {
+    include!("spec.rs");
+}
+use humblegen_rt::hyper;
+use protocol::*;
+use std::sync::Arc;
+
+struct S;
+
+#[derive(Default)]
+struct AuthzScope {
+    user_id: String,
+    posting_allowed: bool,
+}
+
+#[humblegen_rt::async_trait(Sync)]
+impl BlogApi for S {
+    type Context = AuthzScope;
+
+    async fn intercept_handler_pre(
+        &self,
+        req: &hyper::Request<hyper::Body>,
+    ) -> Result<Self::Context, ServiceError> {
+        if req
+            .headers()
+            .get(hyper::header::AUTHORIZATION)
+            .ok_or(ServiceError::Authorization)?
+            .to_str()
+            .map_err(|_| ServiceError::Authorization)?
+            == "Custom AUTHZ_TOKEN"
+        {
+            // .. find user id database ...
+            Ok(AuthzScope {
+                user_id: "alice".to_owned(),
+                posting_allowed: true,
+            })
+        } else {
+            Err(ServiceError::Authorization)
+        }
+    }
+
+    async fn post_user_posts(
+        &self,
+        ctx: Self::Context,
+        post_body: Post,
+        user: String,
+    ) -> Response<Post> {
+        if user != ctx.user_id {
```

Comment:
I know this is example code, but I imagine the API looking something like

```
let user_id = ctx.get_user_id_for_posting()?;
```

The context is opaque beyond that. This prevents a developer forgetting to check auth, because the only way to get the actual user ID is by asking the context (instead of reading it). The function will return a `Result` and respectively a `ServiceError::Authorization` if the user was not logged in.

Reply:
Actually, I would suggest the contrary because we can leverage the `warn(unused)` compiler lint to prevent forgetting thinking about authorization:

```
struct AuthzScope {
  get_user_info_allowed: bool,
}

impl {
type Context = AuthzScope;
fn get_user_info(&self, ctx: Self::Context, ...) -> Response<u32> {
  let AuthzScope{ get_user_info_allowed } = ctx;
 ... // if we forget to do something with `get_user_info_allowed`
     // the compile `warn(unused)` hint will warn us about it.
}
```

Reply:
So we rely on the unused warning for the whole `ctx` here to not forget to deconstruct it? If we're taking something different from the context, there won't by any warning. This seems way more brittle to me, or am I missing something here?

Reply:
> So we rely on the unused warning for the whole ctx here to not forget to deconstruct it?
Exactly, see current HEAD of `cs/develop` in the customer project where we use humblegen.

> If we're taking something different from the context, there won't by any warning. This seems way more brittle to me, or am I missing something here?

Don't see why it is _more_ brittle than a method invocation which doesn't make the slightest squeak if you forget it. 

The project for which we develop humblegen only uses the interceptor feature for auth ATM. If we ever start using it for other purposes we have to rethink the `warn(unused)` trick in the example, but ATM, I believe that it's exactly on point.
One way to keep the pattern would be to have two type variables, one for `AuthzContext` and one for `ReqContext`.

I suggest we defer discussion about this until we have gained more experience _using_ humblegen in the customer project.